File: Frontend/src/app/core/services/productCategory.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ProductCategory } from '../models/productCategory.model';
import { ProductCategoryDTO } from '../models/dto/product-category.dto';

@Injectable({
  providedIn: 'root'
})
export class ProductCategoryService {
  private apiUrl = 'http://localhost:8080/api/product-category'; // Ajusta la URL según tu backend

  constructor(private http: HttpClient) {}

  // Obtener todas las relaciones producto-categoría
  getAllProductCategories(): Observable<ProductCategory[]> {
    return this.http.get<ProductCategory[]>(this.apiUrl);
  }

  // Asignar una categoría a un producto
  assignCategoryToProduct(data: ProductCategoryDTO): Observable<ProductCategory> {
    return this.http.post<ProductCategory>(this.apiUrl, data);
  }


  // Obtener categorías por producto
  getCategoriesByProduct(productId: number): Observable<ProductCategory[]> {
    return this.http.get<ProductCategory[]>(`${this.apiUrl}/categories/${productId}`);
  }

  getProductsByCategory(categoryId: number): Observable<ProductCategory[]> {
    return this.http.get<ProductCategory[]>(`${this.apiUrl}/products/${categoryId}`);
  }

  // Eliminar una relación producto-categoría
  removeCategoryFromProduct(productCategoryId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${productCategoryId}`);
  }
}

File: Backend/src/main/java/com/motionParts/ecommerce/services/ProductCategoryService.java
package com.motionParts.ecommerce.services;

import com.motionParts.ecommerce.dto.ProductCategoryDTO;
import com.motionParts.ecommerce.Models.Product;
import com.motionParts.ecommerce.Models.Category;
import com.motionParts.ecommerce.Models.ProductCategory;
import com.motionParts.ecommerce.repositories.ProductCategoryRepository;
import com.motionParts.ecommerce.repositories.ProductRepository;
import com.motionParts.ecommerce.repositories.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProductCategoryService {

    @Autowired
    private ProductCategoryRepository productCategoryRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    // 🔹 Obtener todas las relaciones Producto - Categoría
    public List<ProductCategory> getAllProductCategories() {
        return productCategoryRepository.findAll();
    }

    // 🔹 Asignar una categoría a un producto con DTO
    public ProductCategory assignCategoryToProduct(ProductCategoryDTO dto) {
        Product product = productRepository.findById(dto.getProductId())
                .orElseThrow(() -> new RuntimeException("Producto no encontrado con ID: " + dto.getProductId()));
        
        Category category = categoryRepository.findById(dto.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Categoría no encontrada con ID: " + dto.getCategoryId()));
    
        ProductCategory productCategory = new ProductCategory(product, category);
        return productCategoryRepository.save(productCategory);
    }

    public List<ProductCategoryDTO> getCategoriesByProduct(Long productId) {
        return productCategoryRepository.findByProductId(productId)
                .stream()
                .map(pc -> new ProductCategoryDTO(pc.getProduct().getId(), pc.getCategory().getId()))
                .collect(Collectors.toList());
    }
    
    public List<ProductCategoryDTO> getProductsByCategory(Long categoryId) {
        return productCategoryRepository.findByCategoryId(categoryId)
                .stream()
                .map(pc -> new ProductCategoryDTO(pc.getProduct().getId(), pc.getCategory().getId()))
                .collect(Collectors.toList());
    }

    // 🔹 Eliminar una relación Producto - Categoría
    public void removeCategoryFromProduct(Long productCategoryId) {
        productCategoryRepository.deleteById(productCategoryId);
    }
}
